MAINT: shim for descr->f access

tylerjereddy opened the pull request:
* Fixes gh-20107

* WIP for now because I don't like that I had to manually adjust the NumPy compat shim header...

* I confirmed that this branch restored full build/test suite passing on x86_64 Linux with NumPy in the following scenarios: build with NumPy `main`, test with NumPy `main`; build with NumPy `main`, test with NumPy `1.26.4` (and 1.x only of course)

* although I did apply the suggestions at or linked from: https://github.com/numpy/numpy/pull/25812
I'm not happy about the fact that I have to modify `npy_2_compat.h` manually (this seems very much not ideal, perhaps I'll open an upstream issue)

[skip cirrus] [skip circle]

tylerjereddy committed:
WIP, MAINT: shim for descr->f access

* Fixes gh-20107

* WIP for now because I don't like that I had to manually
adjust the NumPy compat shim header...

* I confirmed that this branch restored full build/test suite
passing on x86_64 Linux with NumPy in the following scenarios:
build with NumPy `main`, test with NumPy `main`; build with
NumPy `main`, test with NumPy `1.26.4` (and 1.x only of course)

* although I did apply the suggestions at or linked from:
https://github.com/numpy/numpy/pull/25812
I'm not happy about the fact that I have to modify
`npy_2_compat.h` manually (this seems very much not ideal,
perhaps I'll open an upstream issue)

[skip cirrus] [skip circle]

tylerjereddy commented:
Some of the CI failures may be because we use a NumPy 2.x that is older than NumPy `main` at the time of writing (i.e., a pre-release wheel that doesn't have the NumPy C API changes yet). I'm not sure there's any way (or motivation to) make that scenario work as well.

rgommers commented:
> Some of the CI failures may be because we use a NumPy 2.x that is older than NumPy `main` at the time of writing (i.e., a pre-release wheel that doesn't have the NumPy C API changes yet).

Looks like the merge of https://github.com/numpy/numpy/pull/25812 (14 hours ago) just missed the rebuilds at https://anaconda.org/scientific-python-nightly-wheels/numpy/files used in these jobs (27 hours old now).

> I'm not sure there's any way (or motivation to) make that scenario work as well.

Probably not worth it - this will go away as soon as new wheels are uploaded, and I don't think that our CI is failing until that happens? There's only one C API version number, and it got bumped with the PR in question here.

rgommers commented:
I've just started new wheel builds on numpy main, so the problem here should go away within an hour I hope.

rgommers commented:
> I've just started new wheel builds on numpy main, so the problem here should go away within an hour I hope.

Unfortunately, the Windows wheel builds are broken, due to https://github.com/numpy/numpy/issues/25849. So we now have new wheels for Linux/macOS, and the old ones for Windows. So I expect when re-running the failing jobs here that the Linux jobs will pass but the Windows one won't. Let's see.

tylerjereddy commented:
Sounds like Nathan agrees there's some concern that I had to delete some code in the compat header as well, so yeah guess we'll see where the dust settles, just wanted to get it rolling.

rgommers commented:
Looks like this is being reverted in NumPy for now, so let's not merge this PR.

tylerjereddy committed:
MAINT: PR 20108 revisions

* vendor in the latest version of the compat header from:
https://github.com/numpy/numpy/pull/25897

* with this adjustment and that branch, the following
scenarios all have full test suite passing:
  - build + test against his feature branch with new header vendored
into my branch
  - (with new header vendored in) build against his branch, test against 1.26.4
  - (with new header vendored in) build + test against 1.26.4

[skip cirrus] [skip circle]

tylerjereddy commented:
As summarized at https://github.com/numpy/numpy/pull/25897#pullrequestreview-1905018172, things seems "ok" now with Sebastian's latest PR there. I've pushed in a commit here to adjust to the latest version of the compat header.

- build + test against his feature branch with new header vendored
into my branch: full suite passed!
- (with new header vendored in) build against his branch, test against 1.26.4: full suite passed!
- (with new header vendored in) build + test against 1.26.4: full suite passed!

rgommers commented:
Nice, thanks Tyler. That's about as good as it's going to get it seems. We're still going to have broken CI for a little while after the NumPy PR goes in and nightly wheels get uploaded. But we then have this PR ready to unbreak CI.

rgommers commented:
New numpy nightlies are up with the `->f` removal included, so I pushed a rebase. Let's see if CI is happy now.
